src/transformers/pipelines/base.py

```diff
@@ -769,18 +769,41 @@ def __init__(
                 self.device = device
             elif isinstance(device, str):
                 self.device = torch.device(device)
-            elif device < 0:
+            elif device is None or device < 0:
                 self.device = torch.device("cpu")
             else:
                 self.device = torch.device(f"cuda:{device}")
         else:
-            self.device = device
+            self.device = device if device is not None else -1
         self.torch_dtype = torch_dtype
         self.binary_output = binary_output
 
         # Special handling
-        if self.framework == "pt" and self.device.type != "cpu":
-            self.model = self.model.to(self.device)
+        if self.framework == "pt" and device is not None:
+            self.model = self.model.to(device=self.device)
```

Comment:
Don't mix `self.device` and `device` . This is super error prone.

The proposed change I made was at least explicit about it's default value.
I really think this needs to be changed. Too many opportunities to introduce bugs later on.

- Set the default value (if no value provided)
- Handle `device` to create `self.device`.
- Use `self.device` everywhere.

Reply:
Yeah I somehow didn't fully considered your proposition in https://github.com/huggingface/transformers/pull/21479#issuecomment-1420747654 - I think it's wiser to revert my changes with yours!